File: Gombang/[LeetCode75] - [Array_Sort]/238_Product_of_Array_Except_Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums)
    {
        // nums의 길이만큼 vector를 만들어주고 모두 1로 초기화를 해준다.
        vector<int> result(nums.size(), 1);

        // 좌측 곱 계산.
        int multiply = 1;
        for (int i = 0; i < nums.size(); i++)
        {
            result[i] *= multiply;
            multiply *= nums[i];
        }

        // 우측 곱 계산.
        multiply = 1;
        for (int i = nums.size() - 1; i >= 0; i--)
        {
            result[i] *= multiply;
            multiply *= nums[i];
        }

        return result;
    }
};

File: Seori/LeetCode 75/Array_String/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # [1] 왼쪽에서 오른쪽으로 곱한 값을 저장한다. Save the product from left to right.
        length = len(nums)
        answer = [nums[0]] + [0] * (length - 1)
        for i in range(1, length):
            answer[i] = answer[i-1] * nums[i]
        
        # [2] 오른쪽에서 왼쪽으로 곱한 값을 변수 right_product에 저장하면서 answer를 완성한다.
        #     Complete the answer while saving the product from right to left in the variable right_product.
        right_product = 1
        for i in range(length-1, -1, -1):
            # [2-1] 처음과 끝 인덱스 처리
            if i == length-1:
                answer[i] = answer[i-1]
            elif i == 0:
                answer[i] = right_product
            # [2-2] 나머지 인덱스 처리
            else:
                answer[i] = answer[i-1] * right_product
            right_product *= nums[i]
        return answer
        

        
